File: euler.c
#include <stdio.h>
#include <math.h>

float f(float t,float y)
 {
 return y-t*t+1; //y' of the differential equation
 }
int main()
{
 int i;
 float h,t,y,y1,a_e,e_b;
 h=0.2; //stepsize
 t=0; //initial value of t
 y=0.5; //initial value of y
 printf("\n  t\t  y\t    absolute error   error bound\n");
while (t<=2+h)
   {
    y1=(t+1)*(t+1)-0.5*exp(t); //value of y from analytical solution
    a_e=fabs(y-y1); //absolute error in each step
    e_b=fabs(0.1*(0.5*exp(2)-2)*(exp(t)-1)); //error bound in each step
    printf("%0.3f\t%0.3f\t\t%0.3f\t\t%0.3f\n",t,y,a_e,e_b);
    y=y+h*f(t,y); //formula for euler method
    t=t+h;
    }
}
 

File: euler.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Mar 30 22:10:53 2020

@author: suman
"""

import matplotlib.pyplot as plt
import numpy as np

def f(t,y):
    return (y/t)-(y/t)*(y/t) #y' of the differential equation
h=0.1 #stepsize
t_sol=np.array(1) #initial value of t 
y_sol=np.array(1) #initial value of y 
t=1
y=1
while t<2:
    y=y+h*f(t,y) #formula for euler method
    t=t+h
    t_sol=np.append(t_sol,t) #appending the value of t after each step
    y_sol=np.append(y_sol,y) #appending the value of y after each step
plt.plot(t_sol, y_sol,label='solution of differential eq') #plotting of solution
def g(t):
    return t/(1+np.log(t)) #analytical sol of differential eq
x = np.arange(1, 2+h, h)
a_e=np.average(abs(g(x)-y_sol)) #absolute error
r_e=np.average(abs((g(x)-y_sol)/g(x))) #relative error
print('The absolute error is',a_e)
print('The relative error is',r_e)
plt.xlabel('t', fontsize=15)
plt.ylabel('y', fontsize=15)
plt.title('t vs y graph')
plt.legend()
plt.show()
